torch/library.h

```diff
@@ -1,5 +1,63 @@
 #pragma once
 
+/// \file
+///
+/// This header provides an API for extending PyTorch's core library
+/// of operators with user defined operators and data types.  This
+/// API can be used in a few ways:
+///
+/// * You can define new custom operators and classes with TORCH_LIBRARY(),
+///   making them available for use in both eager Python as well as in TorchScript.
+///   This API is modeled off of pybind11's `PYBIND11_MODULE` macro, as
+///   the provided functionality is similar (pybind11 lets you bind C++
+///   to Python only; `torch/library.h` lets you bind C++ simultaneously
+///   to Python and TorchScript).
```

Comment:
A few questions out of curiosity :P :
- it looks like using TORCH_LIBRARY can support both eager and torchscript while TORCH_LIBRARY_IMPL only supports eager by registering to a particular dispatchKey? Is this interpretation correct? 
- If so, does it mean xla can support torchscript by switching to TORCH_LIBRARY? (I might be missing something obvious here...

Reply:
> it looks like using TORCH_LIBRARY can support both eager and torchscript while TORCH_LIBRARY_IMPL only supports eager by registering to a particular dispatchKey? Is this interpretation correct?

Umm, not quite. You can't TORCH_LIBRARY_IMPL without also having a TORCH_LIBRARY. The act of registering a schema for an operator is what makes it available for torchscript and eager. TorchScript will still execute operators through the dispatcher, so that is how you "support" both CPU and CUDA through torchscript. I think XLA could work with torchscript in this way (so you'd still be lazily constructing an xla graph as you execute the python program through the torchscript interpreter), but that's probably not what you mean by torchscript support?